### ISSUE TITLE
Process.spawn fails if project folder isn't a real folder

### ISSUE THREAD

**Issue by rmunn** [ionide/ionide-helpers#3]
I believe I may have found the cause of ionide/ionide-fsi#4 and ionide/ionide-paket#20. In my testing of ionide/ionide-fsi#4, I found that if Atom's current project folder is somewhere that isn't a real on-disk folder, like `atom://about`, then I can't spawn `Fsi.exe`. But if I delete the project folder, or make sure that it's a real on-disk folder instead of a location like `atom://about`, then I no longer get the "(Path) could not be spawned" error.

(To get `atom://about` as a project folder, BTW, simply remove all your existing project folders and then go to Help->About. A new "folder" labeled `about` will show up in the project folders pane; see screenshot below. If you right-click on it and choose "Copy Full Path", then paste that into a text editor, you'll see the path is `atom://about`.)

![atom-spawn-failure](https://cloud.githubusercontent.com/assets/90762/12441953/a1c85c00-bf7e-11e5-8d18-871576e50f5e.png)

It looks like there are several lines of code in https://github.com/ionide/ionide-helpers/blob/master/src/ProcessHelpers.fs where `Globals.atom.project.getPaths()` is used to set the current working directory for spawning processes. The code appears to assume that these will always be real directories, but that is not always the case. If the project folder is `atom://about`, or any other value that isn't a real directory (I could also get the name of a file into there by closing all project folders, then opening an individual file), then that assumption will fail and the `Process.spawn()` function will fail.

Here are the locations in https://github.com/ionide/ionide-helpers/blob/master/src/ProcessHelpers.fs that make this incorrect assumption:

https://github.com/ionide/ionide-helpers/blob/master/src/ProcessHelpers.fs#L73
https://github.com/ionide/ionide-helpers/blob/master/src/ProcessHelpers.fs#L94
https://github.com/ionide/ionide-helpers/blob/master/src/ProcessHelpers.fs#L112
https://github.com/ionide/ionide-helpers/blob/master/src/ProcessHelpers.fs#L118

Also, https://github.com/ionide/ionide-helpers/blob/master/src/SettingsHelpers.fs#L45 should probably be looked at as well.

**Comment by rmunn** [comment]
This also appears to be the cause of ionide/ionide-fsharp#180 -- note the screenshot with an "about" folder in his project-folders pane on the left, and how the Atom window's titlebar contains `atom://about`. And in ionide/ionide-paket#20, note the "config" folder in his project-folders pane, and the Atom window's titlebar that contains `atom://config`.

I suspect this is also the real cause behind ionide/ionide-fsharp#178.

**Comment by rmunn** [comment]
atom/atom#10297 is tracking this issue from the Atom side, so the fake `atom://foo` folders might go away once that bug is fixed. But until it is fixed, ProcessHelpers.fs probably should check for this case.

**Comment by cloudRoutine** [comment]
Great find, we really appreciate the thorough detail of this issue report. I encourage you to make a PR with these changes to record credit for the fix. Otherwise I'll try to get to it in the next day or two, unless @Krzysztof-Cieslak  beats me to it ;)

**Comment by rmunn** [comment]
I don't have a good dev environment set up to test the fix, and I don't really want to make an untested PR. If I compile a new version of ionide-helpers, how would I link that into my running copy of Ionide in Atom? Or would I need to build the complete Ionide project? I've seen https://github.com/ionide/ionide-fsharp/blob/develop/README.md, but I don't know how that would apply to ionide-helpers -- which isn't an apm package *per se*, but a set of DLLs used by the other packages at some point.

Basically, I'd be happy to do it but I need a bit more explanation about the build process: at what point does the F# become Javascript, and how do I get a change in ionide-helpers into my installed Atom package?

**Comment by rmunn** [comment]
And while I'm at it, another change I think should be made: https://github.com/ionide/ionide-helpers/blob/master/src/ProcessHelpers.fs#L85 looks wrong to me. I've verified that if you pass a Javascript `null` as the `options` parameter of Node's `child_process.spawn()`, it will throw a `TypeError`. I think that [ProcessHelpers.fs line 85](https://github.com/ionide/ionide-helpers/blob/master/src/ProcessHelpers.fs#L85):

                | _ -> null |> unbox<AnonymousType598>

should be changed to:

                | _ -> { } |> unbox<AnonymousType598>

As I said in my earlier comment, I haven't been able to test this change in F#. But I did test passing an empty Javascript object to `child_process.spawn()` as its `options` parameter, and it's perfectly happy with that. So in theory, the change above should work.

**Comment by cloudRoutine** [comment]
Currently Ionide Helpers is manually updated across the various projects and stored in the [/lib/](https://github.com/ionide/ionide-fsharp/tree/develop/lib) directory. Eventually we'd like to use github releases as a dependency through paket rather than creating a nuget package for this dl, but that functionality hasn't been added to paket yet.  

Currently the way I test out updates is using this [FAKE Build Target](https://github.com/ionide/ionide-fsharp/blob/develop/build.fsx#L127-L134). You'll need to copy your updated compiled dll into the aforementioned `/lib/` directory before running it.   

The the `Runscript` target (or on Mono `BuildGenerator` and `RunGenerator` targets) create the actual js in the target chain
```
"Clean" ==> "RunScript" ==> "InstallDependencies" ==> "TryPackage"
```
Anything else I can clarify?

**Comment by rmunn** [comment]
That should be enough for me to get started with. But I'm afraid you shouldn't expect any actual code from me for a while; it's getting late here in my time zone, and I need to go home. And then I have several other things on my schedule in the next 2-3 days, so I don't realistically expect to be able to do any actual work on this until next Monday at the earliest. If you want to jump in and fix it before then, please do.

**Comment by rmunn** [comment]
Well, I may need a bit more hand-holding through the build process, as I haven't used FAKE before. My first problem was that Paket wasn't downloading when `build.sh` ran the bootstrapper, but that was solved by running `mozroots --import --sync`. But now I'm getting:

    System.IO.FileNotFoundException: /home/rmunn/code/ionide/ionide-helpers/temp\FunScript.TypeScript.Binding.atom.dll does not exist

when I try to build ionide-helpers. And I'm not sure where to start debugging this problem. `/home/rmunn/code/ionide/ionide-helpers/temp/FunScript.TypeScript.Binding.atom.dll` does exist, so the problem might be that backslash in the path, but I don't know where to look for the erroneous backslash.

**Comment by cloudRoutine** [comment]
how are you running the build script? I dev on windows so I don't know if there's an error in the `build.sh` 
your file tree for the repository should look like 

![](http://i.imgur.com/efkQofq.png)

The issue may have been with path strings for the `.dlls` I just pushed a small edit to them that might resolve the issue 

![](http://i.imgur.com/VoeDL9h.png)

**Comment by rmunn** [comment]
I'm running it by doing `./build.sh` in the root of the repo.

Also, I can confirm that the backslash was the problem. Changing lines 48-52 of `build.fsx` to "temp/FunScript.(etc)" allowed me to successfully build the project. Backslashes as path separators work only on Windows and not on Linux, but forward slashes as path separators work on both Windows and Linux, so forward slashes are better when writing cross-platform code. (Yes, forward slashes as path separators work on Windows; not enough people know that.)

If you like, I'll submit a PR to fix the build scripts on Linux.

**Comment by cloudRoutine** [comment]
we know they work, just look at our other build scripts :wink: but you're right it should be fixed with this one. I had a brain fart earlier and changed the strings to unescaped strings with the `@` and didn't fix the slashes.

**Comment by rmunn** [comment]
Okay, I was able to build this and test it out on my local copy of Atom (on Linux Mint Debian Edition), and it handled the "`atom://about` as project path" scenario just fine.

Do note that I haven't yet tested this on Windows. I'll have access to a Windows machine with Atom in a couple of days, so I can test it then. But I went ahead and made the PR now so you wouldn't have to wait a couple of days to merge it if the code looks good. Just be aware that you might want to test it on Windows.
